### ISSUE TITLE
More clarification on async runtime requirements

### ISSUE THREAD

**Issue by dakom** [Lay3rLabs/WAVS#224]
I'm opening this issue more as like a way to not lose track of some open questions I had following https://github.com/Lay3rLabs/WAVS/blob/main/docs/ASYNC_NOTES.md

### How is it okay to call wstd `block_on` within a tokio `block_on` ?

Here's my current understanding:

1. the wasi side of things does not really have its own async runtime. wstd provides a sort-of runtime api with things like "block_on", but ultimately, the machinery of waking up and driving futures forward is from the host, by way of guests ultimately calling `wasi::io::poll`

2. `wasi::io::poll` is, however, just a WIT definition, at compile time it generates bindings that are ultimately "filled in" by the host - and so in our case it's provided by wasmtime, and our usage of wasmtime gets its runtime via tokio

3. that's why it's okay to call wstd block_on even though it's within the host's tokio block_on... the way wstd block_on works is it's actually just scheduling tasks on the runtime and looping until those tasks are ready - so it is not blocking the the _host_ it's blocking _the component_. Conceptually, it _can_ be thought of as more like a loop with a tokio-aware sleep (this isn't how it actually works - there's a mechanism of polling and waking etc.)

4. practical example: if one component calls wstd block_on, but another component schedules tasks that are able to progress, that other component will be able to move forward

5. why would someone ever use wstd block_on instead of wait_for in a component? no idea, it seems to me that it's ultimately the same as just awaiting... I guess maybe to more clearly manage logic within a component that doesn't want to drive its own internal futures forward, so in that sense it's sortof a real runtime within a runtime, if we squint...

### Do we need separate runtimes per-thread?

I'm actually not 100% sure that we need to give each instance its own runtime... if, say, a component kicks off some http request and then does a whole lot of blocking CPU work, it might be that the CPU work in the component doesn't interfere with the host runtime at all. It'll delay the component's ability to call poll() quickly, so the future won't resolve with a result until the CPU work is done, but this is a component-logic problem, and does not impact the overall health - i.e.  it will not prevent other components from polling the runtime and driving _their_ futures.

In other words, if I'm understanding correctly (and maybe I'm not!) it's actually impossible for a component to block the host async runtime at all. (tangential thought: perhaps this is related to the overall sandboxing design.) At worst it could call poll() too often on a loop, but we don't even hit that problem since the crates we use know how to sleep and wake themselves up when data is ready (or something like that)

Of course, this has nothing at all to do with blocking the host for CPU computation, that can block the host, and so keeping each execution in its own thread is important - it's just that maybe we're fine with one shared tokio runtime everywhere
